### ISSUE TITLE
Issue Handling

### ISSUE THREAD

**Issue by iqualfragile** [minetest/minetest#6394]
recently a lot of issues that the core team does not have the energy to implement themselves have been closed.
this issue is for discussion about how to handle issues in minetest, specifically issues that are seen as a nice idea, but will probably not be implemented in the near or mid future by the core team.

in my opinion the right way to go about such issues would be to tag them with "idea bin" or something along the lines, and just leave them be.

**Comment by jastevenson303** [comment]
It's possible to use the forum for a lot of brainstormy stuff.  IRC, too.  Old closed issues can still be commented on.

If the issue is still worth consideration after all that, here on the tracker, then it can be re-opened.

**Comment by paramat** [comment]
> recently a lot of issues that the core team does not have the energy to implement themselves

Not true, they are closed for a good reason, not because we don't have the energy or time. We are lenient with issues that are a good idea and allow them to sit around for years waiting for someone to get inspired to work on them.

> how to handle issues in minetest, specifically issues that are seen as a nice idea, but will probably not be implemented in the near or mid future by the core team.
On my opinion the right way to go about such issues would be to tag them with "idea bin" or something along the lines, and just leave them be.

Good ideas are not closed, see the oldest issues that are still open since 2012.

**Comment by paramat** [comment]
MT is becoming more popular, and inevitably we have increasing numbers of people casually adding every idea they have to issues, which is essentially fine of course, but leads to the issue number problem. A little while ago we were considering removing feature requests completely from issues, so we are being very lenient in still allowing these. In a way issues could be considered to be for bug fixes and technical issues only, not random ideas for the game.

The reason many issues has been closed recently is because some of us have put in a huge amount of work in going through old issues and closing those which need closing. So isues have closed because of dev attention, not because we have suddenly become much stricter.

Yesterday i spent the whole day going through issues, eventually i was exhausted after 250, imagine what it's like spending so much precious time and effort doing this and knowing there is nearly 1000 issues to review and manage, it's a nightmare. We will have to limit the issue number by any means, by closing old feature requests that don't have much support or interest.

**Comment by paramat** [comment]
I think this is an inherent problem of free open source projects.

'Open source' means every player feels they have the right to make feature requests and suggestions directly to the dev team, and currently we allow these to be added direclty into Github issues. This means a rapid rate of often casually added ideas.

'Free' means a very limited amount of dev time because the work is voluntary. We have 4 or 5 very active devs, but most of these are studying or working full time and have limited time for MT.

I've noticed the rate of issues has been increasing, it wasn't long ago we had 700 issues and c55 was making fun of this number in his 'marketing Minetest' video. Now 700 seems small. We hit roughly 850.
The more issues there are the more overwhelming it becomes which puts devs off processing them, which makes the situation worse. I think a comfortably manageable number would be around 500.

So to keep issues manageable we now have to close them at least as fast as they are opened, this will be difficult and inevitably will offend the issue authors.

A while ago we considered whether feature requests should be allowed in Github issues at all, but we were not keen on the idea. I wasn't keen because sometimes they are specific, practical and useful.

So we are iin a very difficult position where we have to limit the number, but to do so means we have to be tough about issues, which will offend people. Obviously bug reports have to remain, feature requests will tend to be the ones we can close.

Perhaps once a feature request reaches a certain age, 2 years?, we then bump it for attention and be quite tough about closing those with little interest or support.

**Comment by lupoDharkael** [comment]
Here some ideas to help about this problem:
1- If you decide to close a possibly interesting but very old feature request we could have a special tag to make easier to find that kind of requests.

2- Create a new repo for MT-requests and use the issues of the engine's repo for technical/bugs only problems.

3- Define a more restrictive pattern for suggesting features. That would prevent the vague "Hey I just want this implemented in the engine" and would require to explain the feature, why is it needed, what solves etc

4- Propose a more serious use the the Feature Discussion part of the forum.

**Comment by C1ffisme** [comment]
> Perhaps once a feature request reaches a certain age, 1-2 years?, we then bump it for attention and be quite tough about closing those with little interest or support, often the author and supporters may have moved on or changed their opinions.

Should that still apply to feature requests like Dimensions, Mob APIs or Voxel Area Entites that may take years or longer to implement?

Although, maybe a better approach to that is to create a list of "development goals" and larger requests like that will be put in the goals list. Within issues on the topic of Dimensions or other long term goals, devs shouldn't be discussing how to code something, but rather, whether it is possible or not and whether or not it should go in the dev goals list.

A lot of people suggest having suggestions (see what I did there?) in the Feature Discussion forum, but I think a separate forum specifically for Suggestions is called for.

The Suggestions forum should take requests like specific blocks, or things specifically for MTG, or minor API functionality. Github Issues should contain Feature Requests for major API functionality, and the Dev Goals should contain Feature Requests that are certain to make it into the game some day.

A Dev Goals list also means that part of the issue of unclear Development Direction is solved.

**Comment by iqualfragile** [comment]
Maybe adopting an RFC process would be a way to go?
People write detailed proposals that follow some sort of structure (highlevel overview, what problem does it solve, pros/cons, implementation considerations etc) which are then opened for comments until a consensus has been found?

that way existing feature requests can be moved over there, and not weight this tracker down, also the RFC process could run with low dev-invention/needs a different skillset then writing code so maybe new people will step up.

**Comment by paramat** [comment]
Remember closed issues are not deleted, they can still be discussed in the thread.

> If you decide to close a possibly interesting but very old feature request

If an idea has potential it won't be closed.

I wouldn't want to disallow ideas from anyone, someone with no technical knowledge can have a good idea. I think devs need to be more active assessing new requests and state their opinion so that these can be closed quickly if necessary.

**Comment by paramat** [comment]
So, very old feature requests that have little support or interest are the first ones to close.
What i suggest, and what i am doing, is bumping them and asking if anyone capable of coding the feature is interested in doing so at some point, if not, i will close them after a month.
Many are 2-3 years old and if no-one capable of coding it has become interested in this time then i think these can be closed, i wish we didn't have to but if we don't become al ittle more strict we will drown in 1000s of issues.

**Comment by lupoDharkael** [comment]
Newcomers should be able to know this, especially if they plan to contribute with some features.
We should say somewhere that they can check the closed issues tagged as 'possible close' and find some possible ideas to implement (apart from requests in the open issues).
I think this may be added to the CONTRIBUTING.md

**Comment by Wuzzy2** [comment]
You could start by defining what issues NOT to post.
You could also write down what kind of features are no-go areas and will never be added.

Issues should also be decided on more quickly and pragmatically. If someone posts a PR with a tiny change which changes one or two sentences, usually there will be immediately 20 comments commenting on the exact choice of words, while the huge API changing issues have 2 comments at most. This is known as “bikeshedding” and very annoying and a complete waste of everyone's time. Discussions like these have to die quickly, please don't let them get out of hand.
Issues which have been discussed to death over very minor details should be simple dealt with and closed more quickly, so that developer attention can go on with more important issues.

It would also help a lot if a general development / project guideline is written, so people know where Minetest will be heading to overall. And also state what is NOT a goal of Minetest (very important!). This will hopefully prevent some of the endless discussions as the key questions are already answered.

The reason why there are so many suggestions is because Minetest is still very WIP and in alpha stage and is, frankly, *lacking* in features. Keeping feature requests in GitHub makes sense.

> This means a rapid rate of often casually added ideas.

Maybe *casual* ideas (ideas with little thought or consideration for the project) should be rejected and closed quickly and more rigerously. As you already said, there *are* good suggestions, but there are also bad ones. The bad ones are suggestions with no rationale, with poor explanation, with lots of room open for speculation, etc. I think you could do a policy in that suggestions must be stated clearly, be well-defined and not just some vague hogwash with little thought. You could then go on close the garbage issues much quicker without guilt. ;-)

Another possibility is to close all issues which are neither bug report nor **specific** feature requets immediately (with the reason: “off topic”) and refer to the forums or IRC for general discussion.

In other words, issues like this one! :D 
Let's continue discussion in the forums, shall we? ;-)

**Comment by Wuzzy2** [comment]
Forget my previous post. When I think about it, the number of *justified* issues will continue to rise, no matter what. Closing unjustified ones simply won't do the cut, it's not going to help a lot. We need a complete new way of handling issues.

Actually, a very smart person came up with a very clever solution for the exact problem you're having:
http://www.lostgarden.com/2008/05/improving-bug-triage-with-user-pain.html

**Comment by C1ffisme** [comment]
While a closed discussion can still be talked in, it's easy for it to get lost. So even if a discussion like Dimensions, Mob APIs, VAEs or else has a lot of support, but is closed after 2-3 years to help sort feature requests from bug reports and such, it will slowly die off as people's notifications are filled with bug reports that are open, and people start ignoring the closed ones.

Then the feature requests only get resurrected when someone posts a duplicate request and it gets closed, and the main one linked. The post stays alive for a while, but no work will get done.

I still think a list of decided-upon goals needs to be added to the Minetest repository in a Markdown file (called something like `Feature Requests.md`), which would look something like:

```
We accept feature requests as issues here! However, you may
notice that some requests that have not been implemented are
closed. This is NOT NECESSARILY because we have chosen
to ignore them. You can still talk in a closed issue, and we want
to keep urgent bug reports away from the old suggestions.

There are some feature requests that we have planned on
adding to the game, and we've chosen to list them here along
with their issue numbers for discussion. That way we can
keep the open issues page for bug reports, and still be able
to talk about adding certain Features!

Features we plan on adding:
<A list of Issue names with their respective pages linked>

There are some requests we may have closed because we do
not think should be added because they are not possible without 
lowering performance, they will take up too much developer time,
or other reasons.

Features we do not plan on adding:
<A list of Issue names with their respective pages linked>
```

This makes it:
* Easy to spot duplicate requests and link the correct ones AND
* Easy to get an idea of what has yet to be done, and _somewhat_ of a Development Direction.

**Comment by paramat** [comment]
Mob API and VAE issues will not be closed as those have a lot of developer support and are stated in roadmaps, they won't be forgotten.

My suggested approach is this:

Old feature requests (let's say older than 2 years) that are not high priority and have no stated coder interest are labelled 'possible close' and are bumped for attention with a message that they will be closed in 1 month if no coder states an intent to code it in the near-future.

I am starting this already and it is already a little successful in getting some requests coded or on coders TODO lists.
It's fairly strict because i no longer feel that player interest is enough to keep an old request open indefinitely, otherwise we have no hope of controlling the number of requests, any idea will always have some people who like it but if no-one will code or state an intent to code it what's the point?

If in 2 years no coder has stated an interest in working on it then it's likely no-one will in the next 2 years, the request needs attention and a little pressure to either get some commitment or the disapprovals needed to close it.
Obviously there are no hard rules and each case has to be considered individually with many factors taken into account.

This is very much a dilemma situation where we obviously would like to keep all requests open, but being this lenient is now becoming a big problem and a large number of issues will mean fewer getting any attention. Keeping everything open will not help those who open issues.

**Comment by C1ffisme** [comment]
Just as a thought, it might be useful to have certain members of the development community sign up for certain positions in Minetest. (With their consent, of course. They will not be forced to take up a certain role and be forced into hard labor.)

Example positions being:
- Graphical Designer
- Storywriter (Not an essential position now, but could be useful for the small bits of lore in MTG.)
- Mapgen Designer
- Code Design Inspector
- Creature Designer/Mob Behaviorist
- etc.

If people sign up for these roles, then they will be assigned to certain issues automatically (And by automatically, I mean by @paramat or someone else in charge will assign them.) and can be in charge of dealing with the issue.

I'm not saying that what I've spilled out here is a complete system for dealing with issues, but it's worth thinking about. 

(I imagine it'll give some people a headache though, because some areas like the Rendering Engine or the Networking will have lots of issues compared to something like Storywriting. And honestly, you'll have a lot of people wanting to get the more fun jobs like Storywriting and Graphical Design, considering a lot of the newer members joining onto the project have variable skill with C++, Lua or any programming language at all.)

**Comment by iqualfragile** [comment]
what i do not fully understand yet is why a big number of issues is a problem, wouldn't it be fine to just tag them once they are created (or go through old issues and tag those) so people/devs can just look at the categories they are interested in?

i mean an issues tracker is there for tracking issues after all, so i would think that closing an issue means "this is not an issue" and not "ok this is an idea, but we are probably not going to implement it"

if this issue tracker is supposed to be only a bug tracker then yes, a lot of things would need to be closed, but historically this is not how this was used, so yeah.

**Comment by iqualfragile** [comment]
what i do not fully understand yet is why a big number of issues is a problem, wouldn't it be fine to just tag them once they are created (or go through old issues and tag those) so people/devs can just look at the categories they are interested in?

i mean an issues tracker is there for tracking issues after all, so i would think that closing an issue means "this is not an issue" and not "ok this is an idea, but we are probably not going to implement it"

if this issue tracker is supposed to be only a bug tracker then yes, a lot of things would need to be closed, but historically this is not how this was used, so yeah.

**Comment by Wuzzy2** [comment]
I have the feeling that this comment fell completely under the table, so I just repeat it here:

> Actually, a very smart person came up with a very clever solution for the exact problem you're having:
> http://www.lostgarden.com/2008/05/improving-bug-triage-with-user-pain.html

No comments on this?

@C1ffisme: I don't see a point in assigning people, it seems like a complete waste of time. If someone wants to start fixing a bug or adding a feature, that someone usually just starts coding. For bigger complex problems, people can simply write “I started to work on this.” to avoid collisions. Very simple.

**Comment by iqualfragile** [comment]
this is not the exact problem thats present here: in the scenario presented at lost garden there is a fixed amount of time that needs to be distributed to bugs.

but exactly 0 people are getting paid to work on minetest for some time, people work on things they care about, so how much time gets spend on bugs does not depend on user pain but developer interest, and you can't force it any other way.

so the right way to go to get more/more important bugs fixed is to somehow get more developer interest, for example by tagging easy to solve issues and offering mentoring on those to interested new developers, etc.

but this issue is mainly about finding a good way to handle (large amounts of) issues

**Comment by nerzhul** [comment]
For an interesting repository with massive PR & issue, please look at https://github.com/ansible/ansible , our situation is not crazy :) for me only PR should decrease

**Comment by rubenwardy** [comment]
decreasing numbers of PRs is way more important. I was only for closing definitely not very useful issues, and only because of them not being very wanted/popular - not due to lack of 'coders'.

**Comment by C1ffisme** [comment]
@Wuzzy2 

> but exactly 0 people are getting paid to work on minetest for some time, people work on things they care about, so how much time gets spend on bugs does not depend on user pain but developer interest, and you can't force it any other way.

Pretty much this. The reason I made it so that you have to give your consent to be put in a role is because once you get assigned to something, that should be a priority. (Obviously we can't force someone to work over the internet, and paying people isn't a good option for a project with no profit, it's kind-of an honor system thing.)

The other reason is that certain people can become experts on areas of code (Mapgen, Rendering, etc.) and when an issue appears in that area, people will know who to call on.

**Comment by iqualfragile** [comment]
this is not the exact problem thats present here: in the scenario presented at lost garden there is a fixed amount of time that needs to be distributed to bugs.

but exactly 0 people are getting paid to work on minetest for some time, people work on things they care about, so how much time gets spend on bugs does not depend on user pain but developer interest, and you can't force it any other way.

so the right way to go to get more/more important bugs fixed is to somehow get more developer interest, for example by tagging easy to solve issues and offering mentoring on those to interested new developers, etc.

but this issue is mainly about finding a good way to handle (large amounts of) issues

**Comment by iqualfragile** [comment]
(im sorry for the double posts, bad internet :()

**Comment by raymoo** [comment]
Problem with closing issues just because nobody currently wants to work on them is that there is no way currently to differentiate between issues closed because they were implemented or were a bad idea, and issues that are perfectly fine and desirable ideas but during some 1-month period did not have someone willing to program. What if somebody is looking for issues to adopt? If you are going to close issues like this, please at least add a label indicating that it was due to a lack of programmer interest.

**Comment by Wuzzy2** [comment]
User Pain might still help for priotirizing bugs and make more urgent bugs more visible to everyone, while less urgent bugs drop to the bottom. Currently, all bugs are just thrown into a huge bucket, sorted by entry date by default and there are only two priority labels which are decided on a completely arbitrary basis (whereas in User Pain, the priority levels are well-defined). That's not very organized. Closing issues not based on merit makes this problem worse, as raymoo just perfectly explained.

You can say Minetest is voluntary and you're 100% right but you're missing the point. The goal here should simply be to make the number of issues more managable. That way, everybody benefits.

I also think GitHub's issue tracker is really bad, especially for large projects like Minetest. Minetest has come at a point that the GitHub issue tracker really shows its own limitations.

I suppose it's not actually the large number of issues which is the problem but that the existing issues are poorly categorized in terms of importance. User Pain might be one way, but if you can think of a better one, I'm excited to hear about it.


About PRs:
I totally agree that open PRs are actually a much bigger problem. Our current PR process has a huge tendency for contributor burnout, only the most determined “survive”.
I hate especially that pesky “rebase needed” label. I hate rebasing because I have to risk my repository every time I use it; if I do one terrible mistake, I destroy my entire repository. Yet in Minetest it's common practice to request this. Sometimes the same PR can have a “Rebase needed” even twice!
“Rebase needed” is also frustrating if the PR perfectly worked since day 1, was 100% compatible with master, but had no reaction for a long time, then it got finally reviewed and now the code base has changed so much that large portions of code must be thrown away and rewritten. OK, it's rarely so extreme but it is common that enough time has passed that a rebase is often needed for a PR unless it has been merged very quickly. This generates a lot of overhead work for the contributor which could have been avoided, so contributors will quickly percieve it as pointless and a waste of time. I don't think many contributors have the stamina to endure this process.

**Comment by raymoo** [comment]
I feel less nervous about rebasing but I agree that it's annoying to be asked to do it and then never get any reviews.

**Comment by paramat** [comment]
> it might be useful to have certain members of the development community sign up for certain positions in Minetest

We used to do this but stopped, many devs are multi-talented and work on many parts of code, assignment doesn't help at all, it may make devs think 'this isn't my responsibility so i won't work on it'. Also of course, most devs want to work on many various areas of code, and this should be encouraged.

> I have the feeling that this comment fell completely under the table, so I just repeat it here:

I started to read it but when i realised how complex their system is i realised no-one will have the time or energy to use a complex system for categorising bug priority, and we already have 3-level  priority labels that are good enough. Also, it's not bugs that are the problem, it's feature requests.

> there is no way currently to differentiate between issues closed because they were implemented or were a bad idea, and issues that are perfectly fine and desirable ideas but during some 1-month period did not have someone willing to program.

I'll just add that this will be closing because no coder has stated an interest in 2 years plus 1 month, not just 1 month.

I must say though that it's a dilemma and i'm still unsure of what to do. I'm unsure about the 'close in 1 month' part. I would much rather devs actually look at low priority feature requests and disapprove them as being not worth the effort to get them closed.

But i notice many devs don't seem to add their opinions, i think i do more than anyone, which is a little uncomfortable for me because often many devs disapprove of something but i'm the one expressing it, so i get the flak for it :) It also means i get a reputation of being highly opinionated when i'm only doing what all devs should do :) I have become quite unpopular because i comment so much, i don't want to be the scapegoat for the unpopular decisions we have to make :) I'm actually a shy introverted person who hates being a focus of attention.

Please will all devs go through issues and add their opinions? There are many closable requests that just need some disapproval.

**Comment by C1ffisme** [comment]
@paramat

> assignment doesn't help at all, it may make devs think 'this isn't my responsibility so i won't work on it'. 

Ah, I see. I was hoping this would have the adverse effect ("I will make it my responsibility to maintain these sections of code and deal with the issues that come from it."), but I guess not.

**Comment by paramat** [comment]
Well, maybe it would help in some ways. But our new more flexible system seems an improvement.
I haven't noticed any speed up or slow down in dev over that change, that tends to be determined by the amount of dev time instead of anything else.

**Comment by paramat** [comment]
So, i'm going to remove the 'close in 1 month' thing because it doesn't feel quite right to me.
But i am still requesting opinions, closes or commitments from core devs and i still feel we have to attend to issues more and be a little more strict than we have been.

**Comment by iqualfragile** [comment]
@paramat (or the dev team) it seems to me that the primary issue here is not that there are many issues, but that there are too many issues to keep in your head at the same time.

i think that your approach of structuring them into no activity/activity by closing the issues that have dev support is correct. if things get too large you need to structure them, thats true for data structures as well as our brains. my request is to use more labels for structuring. sit down as a group and talk about which things you would like to search for and create those labels. You can also bookmark github searches, just as a sidenote. i would strongly recommend to also add the two labels "easy" and "tutoring available" to lure in new devs.

that way everybody would be happy, things that are not resolved or invalid stay open as an issue, and tagging old issues is not more work then closing them, as the main work is reading and understanding the issue and discussion, not clicking the close-button. Also it probably helps even more then simply reducing the number of issues.

**Comment by lupoDharkael** [comment]
@iqualfragile we already have the "trivial" tag, and if someone wants help about implementing something I suppose the minetest-dev IRC chat is available for everyone.

**Comment by Wuzzy2** [comment]
If an issue is closed with the only reason “lack of developer interest”, then *please* add a special label for this. This is quite different than invalid issues (e.g. “works for me”).
Otherwise, good ideas will be very hard to recover.

**Comment by paramat** [comment]
> this issue is for discussion about how to handle issues in minetest, specifically issues that are seen as a nice idea, but will probably not be implemented in the near or mid future by the core team.

> If an issue is closed with the only reason “lack of developer interest”, then please add a special label for this.

We have been considering this recently. The fact is that many feature requests go years with no dev interest or support. Previously we very kindly left these open, however this means that issue number will always increase and increasingly consist of requests with no dev interest.

In the last day or so we considered that old feature requests that have had no or little dev interest for a long time (2-3 years?) should be closed. I added a label for these so they could be searched for but that label has been deleted now :) I'm not too bothered there wasn't much need for it.

**Comment by paramat** [comment]
> Otherwise, good ideas will be very hard to recover.

If core devs are not interested over 2-3 years it's probably not a good idea and not worth keeping open. The 'won't add' label is probably suitable for these issues.
